### ISSUE TITLE
FFD fails in JModelica tests due to unsupported OS

### ISSUE THREAD

**Issue by mwetter** [lbl-srg/modelica-buildings#612]
The test
`Buildings.ThermalZones.Detailed.Examples.FFD.NaturalConvectionWithControl` as well as all other FFD models fail in the JModelica tests with
```
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning]   <ModelicaError category="warning"><value name="msg">"Error: Unsupported operating system in cfdStartCosimulation.c.&#10;"</value></ModelicaError>
```

**Comment by mwetter** [comment]
Merged to verification test branch for further testing.

**Comment by mwetter** [comment]
This still fails in the Jenkins tests with
```
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning]   <ModelicaError category="warning"><value name="msg">"Error: Unsupported operating system in cfdStartCosimulation.c.&#10;"</value></ModelicaError>

```

**Comment by tsnouidui** [comment]
@agnesramle @chria
I think I have added the compiler directive which works with MinGW (@mwetter will approve the pull request to include it in the verification branch). However I am now getting following error messages when I try to run the model locally:

```
FMIL: module = FMICAPI, log level = 5: Calling fmiInitialize
FMIL: module = Model, log level = 4: [INFO][FMU status:OK] <JMIRuntime><value name="build_date">"Feb  1 2017"</value> <value name="build_time">"15:39:37"</value></JMIRuntime>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning] <ModelicaMessage category="warning"><value name="msg">"... loading ""tab1"" from ""Z:/Ubuntu/proj/buildings_library/models/modelica/git/buildings/modelica-buildings/Buildings/Resources/weatherdata/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.mos""&#10;"</value></ModelicaMessage>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning] <ModelicaMessage category="warning"><value name="msg">"... loading ""tab1"" from ""Z:/Ubuntu/proj/buildings_library/models/modelica/git/buildings/modelica-buildings/Buildings/Resources/weatherdata/USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.mos""&#10;"</value></ModelicaMessage>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning]   <UBoundExceeded category="warning"><value name="message">"Out of bounds for variable roo.irRadExc.J[1]"</value></UBoundExceeded>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning]   <UBoundExceeded category="warning"><value name="message">"Out of bounds for variable roo.irRadExc.J[2]"</value></UBoundExceeded>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning]   <UBoundExceeded category="warning"><value name="message">"Out of bounds for variable roo.irRadExc.J[3]"</value></UBoundExceeded>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning]   <UBoundExceeded category="warning"><value name="message">"Out of bounds for variable roo.irRadExc.J[4]"</value></UBoundExceeded>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning]   <UBoundExceeded category="warning"><value name="message">"Out of bounds for variable roo.irRadExc.J[5]"</value></UBoundExceeded>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning]   <UBoundExceeded category="warning"><value name="message">"Out of bounds for variable roo.irRadExc.J[6]"</value></UBoundExceeded>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning] <ModelicaMessage category="warning"><value name="msg">"Start cosimulation&#10;"</value></ModelicaMessage>
FMIL: module = Model, log level = 3: [WARNING][FMU status:Warning] <ModelicaError category="warning"><value name="msg">"Error: Could not find dll handle.&#10;"</value></ModelicaError>
FMIL: module = Model, log level = 2: [INFO][FMU status:Error] Initialization failed.
FMIL: module = FMICAPI, log level = 5: Calling fmiFreeModelInstance
FMIL: module = FMILIB, log level = 5: Releasing FMU CAPI interface
FMIL: module = FMICAPI, log level = 5: Successfully unloaded FMU binary
FMIL: module = FMILIB, log level = 5: Releasing allocated library resources
FMIL: module = JMPRT, log level = 5: Removing c:\users\thierry\appdata\local\temp\JModelica.org
```
One issue that I see is that Line 153/155 of this code https://github.com/lbl-srg/modelica-buildings/blob/issue612_ffd_mingw/Buildings/Resources/C-Sources/cfdStartCosimulation.c
loads a library which is supposed to be in the `Resources` folder. Now how does this work with JModelica which exports the model as an FMU? Will PyFMI be able to resolve this path at runtime?

**Comment by jespermattsson** [comment]
@tsnouidui
I looked at the C code you linked, and it looks like it depends on the current directory being the root of the library. Dymola (and presumably OpenModelica) does that when you load the library. For this to work in JModelica.org you have to explicitly set the current working directory before simulating, since we do not change it.

The current behavior is likely somewhat fragile in Dymola and OpenModelica as well. You could for instance change the current working directory from within the tool after loading the library, or create an FMU and then restart the tool before loading it (without loading the Building library after restart). Both of those scenarios should lead to the same behavior.

There are some options. The simplest one is probably to use the linker to connect to the DLL instead of doing LoadLibrary in the code. This can be accomplished by simply adding a Library annotation to the external function. `Library="ffd"` should be enough for this case, since the library is in the default location for libraries. If you do that, then JModelica.org will automatically select the correct platform and include the DLL in the FMU.

If the C code for some reason requires the use of LoadLibrary rather than linking against the DLL, then it would be possible to use Modelica.Utilities.Files.loadResources with a "modelica://" URI pointing to the library, and pass that string to the external function, that in turn passes that to LoadLibrary. That would also lead to JModelica.org automatically including the DLL inside the FMU. The downside is that you then need the logic that selects the platform in Modelica code, so I do not recommend this option if the first one is possible.

**Comment by tsnouidui** [comment]
Thanks @jespermattsson. Option 1  seems indeed to be the simplest but I think that the code needs the LoadLibrary since it uses afterwards the handle return by it on lines 181 and 193. I think we will have to go with Option 2 unless you have another suggestion. I am cc'ing the original developers @zuowangda @tianwei1989  so they confirm that LoadLibrary is indeed needed here. @mwetter FYI.

**Comment by tianwei1989** [comment]
@tsnouidui @zuowangda @jespermattsson I  believe that the implicit linking of DLL can work once being provided the lib in compiling. The lines Thierry mentioned are for finding the function and transferring the memory address. I personally preferred the second route as we have invested lots of time in the explicit linking of DLL in this case. However, if the aforementioned second route cannot work out I would be more than happy to try the first one.

**Comment by jespermattsson** [comment]
@tianwei1989 @tsnouidui @zuowangda
If possible, going for the first option will probably be more compatible with other tools, since it is what the Modelica specification describes for linking DLLs. It might be tricky to get the second one to work in Dymola, JModelica.org and OpenModelica all at the same time.

**Comment by tianwei1989** [comment]
@zuowangda @jespermattsson In that case I can try the first method in Windows OS first. @tsnouidui could you please create a branch in the repository for me to work on this issue?

**Comment by tsnouidui** [comment]
@zuowangda @tianwei1989 I created a new branch (issue612_ffd_wtian) for the implementation of Option 1. 
See https://github.com/lbl-srg/modelica-buildings/tree/issue612_ffd_wtian

**Comment by tianwei1989** [comment]
Thanks @tsnouidui for creating the branch. I will work on the issue in the coming weekends. I will keep you guys posted. @zuowangda

**Comment by tianwei1989** [comment]
@tsnouidui I reckon that I may work on Option 1 initially in Windows OS . Then I will send you a pull request to merge into repository such that you can check if the aforementioned problem can be solved or not. Afterwards, if the solution is confirmed workable, then I can continue working on compile file in Linux system. Any comment? @zuowangda

**Comment by tsnouidui** [comment]
Sounds good.

**Comment by tianwei1989** [comment]
@tsnouidui I have sent you the pull request as I have successfully tested implicit link of DLL through lib. There are several problems I noticed:
1. As you can see in my pull request, I have to put the ffd.dll at the same folder level of the executable, namely, Buildings. Otherwise, the program will complain of missing of "ffd.dll".
2. In the file "cfdCosimulation.h", the "modelica_ffd_common.h" from `Resources/src/FastFluidDynamics` is included. In the future FMU work, could this be a potential risk?
Please let me know if you have any comment.
@zuowangda @jespermattsson

**Comment by tsnouidui** [comment]
@tianwei1989 @zuowangda When I try to compile it in JModelica I am getting an 

> undefined reference to `ffd_dll` in `cfdStartCosimulation.c` error

Do you know what is going on? Will it possible for you to to get JModelica and try to translate the model on your machine so you can reproduce the problem?

By the way, Don't use `//` for comment since it isn't supported by the current compiler settings

**Comment by tianwei1989** [comment]
@tsnouidui I will test the codes in JModelica. @zuowangda

**Comment by tianwei1989** [comment]
@tsnouidui while I am installing JModelica on my Windows machine, I am searching around to resolve the problem you reported. What I may suggested is to put an `extern` before `int ffd_dll(CosimulationData *cosim);` in `cfdCosimulation.h` as (I believe) this can tell the compiler that ffd_dll is indeed defined but not shown up in the files in `C-Sources`.  @zuowangda

**Comment by tsnouidui** [comment]
@tianwei1989 Did you ever make this work on your machine? I just tried with version 2.0 of JModelica but the compilation fails.

**Comment by tianwei1989** [comment]
@tsnouidui Unfortunately I haven't successfully installed the JModelica on my Windows machine. Therefore, I did not ever make it work on my computer. Maybe I should try the Linux system as I found a docker that may help to install the JModelica.

**Comment by tsnouidui** [comment]
Try to install the 2.0 Windows version. I could then help you set up things so you can run the model.

**Comment by tsnouidui** [comment]
@tianwei1989 I went back and tested that code with Dymola 2017 FD01 but it doesn't work there either. Did you run https://github.com/lbl-srg/modelica-buildings/tree/issue612_ffd_wtian with Dymola and it worked?

**Comment by tsnouidui** [comment]
I did further investigation and found out that I had to add `Library="ffd"` in the annotation of `Buildings.ThermalZones.Detailed.BaseClasses.cfdStartCosimulation` so the model can compile.

After the change, I tried to run the model but it failed. 
I changed the working directory to be `Buildings` and then the model could simulate. 

Will the current implementation only works if the working directory is `Buildings`? This seems to be the same behavior as before not?

**Comment by tianwei1989** [comment]
@tsnouidui I believe the model can run on my Windows computer. As I have no access to my workstation right now, I can get back to you later once I can remotely control my computer.

**Comment by mwetter** [comment]
`Utilities/IO/Python27/Functions/BaseClasses/exchange.mo` uses the code below
```
    annotation (Library={"ModelicaBuildingsPython2.7",  "python2.7"},
      LibraryDirectory={"modelica://Buildings/Resources/Library"},
      IncludeDirectory="modelica://Buildings/Resources/C-Sources",
      Include="#include \"python27Wrapper.c\"",
      __iti_dll = "ITI_ModelicaBuildingsPython2.7.dll",
      __iti_dllNoExport = true);
```
The FFD program also uses a compiled library, but I do not see where the code contains information for where the library is. Is this the root cause of the problem?

**Comment by tsnouidui** [comment]
@mwetter This is I believe missing on the repo I had to locallly extend the annotation of 
```
external"C" retVal = cfdStartCosimulation(
    cfdFilNam,
    name,
    A,
    til,
    bouCon,
    nPorts,
    portName,
    haveSensor,
    sensorName,
    haveShade,
    nSur,
    nSen,
    nConExtWin,
    nXi,
    nC,
    rho_start)
```
in `Buildings.ThermalZones.Detailed.BaseClasses.cfdStartCosimulation`
to include `LibraryDirectory="modelica://Buildings/Resources/Library", Library="ffd"`.
This was necessary so the model can even compile. However the model will only run if the working directory is `Buildings`?

**Comment by mwetter** [comment]
@tsnouidui Is this also the case if the MODELICAPATH is set properly? It is odd that Python seems to work, but not the FFD code.

**Comment by tsnouidui** [comment]
I tried and set the `MODELICAPATH` to `modelica-buildings` but got the same error. Maybe there is something hard-coded in the` ffd.dll `which is still causing the code to search for the `DLL` at a location which is relative to `Buildings`

**Comment by tsnouidui** [comment]
@tianwei1989 I think. First will be to determine why we need to run the code with `Buildings` as working directory in Dymola. I couldn't find any hard-coded path so I am not sure where the problem is.

**Comment by tsnouidui** [comment]
@tianwei1989 Any updates on this? Were you able to reproduce the issue on Windows with the requirement of having the working folder being `Buildings`?

**Comment by tianwei1989** [comment]
@tsnouidui I just tried on my workstation with a Windows system. When I run the simulation, I need to set the work path to the `Buildings`, otherwise, the model will not run. Previously, there are several users reported the issue, as we explicitly loaded the DLL. I guess if the reason could be that the Modelica environment (in my case is Dymola) cannot find the Lib or DLL, located in the 'Resources' folder?

**Comment by tianwei1989** [comment]
@tsnouidui I not sure how to realize it, but can we set the `buildings` path in to Dymola search path, once we execute the coupled simulation?Codes as follows in `cosimulation.h` need to find `ffd.lib`.
`#pragma comment(lib, "ffd.lib")`
If not, then the model won't run.

**Comment by tianwei1989** [comment]
@tsnouidui Just for your information, the [path ](https://groups.google.com/forum/#!topic/modelica-buildings/c3B86q7gG8E)is a post in Modelica buildings group, which is a matter of working path for Dymola.

**Comment by tsnouidui** [comment]
I am not certain whether we should do it and also how this will be applicable to JModelica?

I still don't have the answer to following question: Why does the code work when we run it with `Buildings` being the working directory? 

In the past we could understand it since the path to the` dll `was assuming that we were using `Buildings` as the working directory. 

But why does the current approach with implicit linking still fail if `Buildings` is not the working directory?

**Comment by tianwei1989** [comment]
@tsnouidui My understanding is that `#pragma comment(lib, "ffd.lib")` in `cfdCosimulation.h` may also assume `Buildings ` as the working directory.

**Comment by tsnouidui** [comment]
`#pragma comment(lib, "ffd.lib")` has zero impact on the code. I commented it out and was still able to run the code.

**Comment by tianwei1989** [comment]
@tsnouidui I just comment it out and I found Dymola threw out error in translation, when using the `implicit link`.

**Comment by tsnouidui** [comment]
what error did you get?

**Comment by tsnouidui** [comment]
Did you change  Buildings.ThermalZones.Detailed.BaseClasses.cfdStartCosimulation

```
external"C" retVal = cfdStartCosimulation(
    cfdFilNam,
    name,
    A,
    til,
    bouCon,
    nPorts,
    portName,
    haveSensor,
    sensorName,
    haveShade,
    nSur,
    nSen,
    nConExtWin,
    nXi,
    nC,
    rho_start)
```
to include LibraryDirectory="modelica://Buildings/Resources/Library", Library="ffd".

**Comment by tianwei1989** [comment]
@tsnouidui Once I added that into the Buildings.ThermalZones.Detailed.BaseClasses.cfdStartCosimulation, the model can run after commenting out the `#pragma comment(lib, "ffd.lib")`.

**Comment by tianwei1989** [comment]
@tsnouidui please hold on as I believe I found some clues.

**Comment by tianwei1989** [comment]
@tsnouidui The reason lays in the '.ffd' files in `Resources/Data/Rooms/FFD`. If you open any `.ffd`file, let's say `ForcedConvection.ffd`, line 6 and line 7 are files to input geometry information. The relative path to them can only be valid if assuming `Buildings` as working directory.

A solution to this issue, as I can envision, is to directly specify the TWO pathes in '.ffd' from the Modelica model.

**Comment by tsnouidui** [comment]
Good catch. That seems to be the problem. 

As you wrote, a potential solution will be to specify the path of the` .cfd` and the `.dat `files in the `.mo` file. This will definitely make the code less fragile as relative paths are always tricky.

Could you create a branch with this one as the base and then make changes there?

@mwetter @zuowangda  It will be good to have your feedback on this design as this will lead to changes to some Modelica models and documentation such as Buildings.ThermalZones.Detailed.CFD.

**Comment by jespermattsson** [comment]
If you specify the paths as "modelica://" URIs and pass them through Modelica.Utilities.loadResource at the call site of the external function, then it will behave well in both JModelica.org and Dymola. In JModelica.org, the files will be included in the FMU, so it should be portable.

Example:
`parameter String dat_uri = "modelica://Buildings/Resourses/some/path/somefile.dat";
// ...
the_external_function(Modelica.Utilities.loadResource(dat_uri), some, other, arguments);`

The reason that you need to put the call to loadResource at the call site is that String parameters isn't fully supported, and are calculated at compile time. So using a parameter to store the result of loadResource would result in the absolute path to the original file location would be used in the FMU, so that it wouldn't be portable.

**Comment by tianwei1989** [comment]
@tsnouidui Sure, will do after @zuowangda @mwetter express their thoughts on this issue. Just to clarify, you want me to create a branch based on `issue612_ffd_wtian`, right?

**Comment by tsnouidui** [comment]
Yes. Since I think that we still want to link the dll implicitly. I thought we could separate both issues (dll and path) but maybe we could also just work on this one if that makes sense?

**Comment by tsnouidui** [comment]
@jespermattsson Thanks for the suggestion

**Comment by tianwei1989** [comment]
@tsnouidui Both work for me. I personally prefer to work on the same branch as the issue of changing path is derived from implicitly linking the DLL.

**Comment by tsnouidui** [comment]
OK. Then let's do it on the same one.

**Comment by tsnouidui** [comment]
@tianwei1989 I suggest to start with the implementation.

**Comment by tianwei1989** [comment]
@tsnouidui will do.

**Comment by tianwei1989** [comment]
@tsnouidui I have completed my implementation. As my adviser just moved to new university, currently I have no Dymola license to test. Once the license is available, I will complete the test and send you a pull request.

**Comment by tsnouidui** [comment]
@tianwei1989 Thanks. I will test the implementation on my side too.

**Comment by tsnouidui** [comment]
@tianwei1989 I wanted to test your code but don't know where it is currently hosted? On which branch are you working on?

**Comment by tsnouidui** [comment]
@tianwei1989 Just quickly checking on this. I could test your code if I knew where to get it?

**Comment by tsnouidui** [comment]
@zuowangda any information about @tianwei1989 code?

**Comment by tsnouidui** [comment]
@zuowangda JModelica can compile and simulate some of the FFD models
https://jenkins.jmodelica.org/job/JM-trunk_Buildings-trunk_verify/lastSuccessfulBuild/artifact/diag/index.html. However, some can't be simulated. There seems to be an issue with threads.  When I try to simulate an FFD example twice in a row in JModelica, then the first simulation succeeds, whereas the second simulation fails. I looked at bit at the code and saw that a thread will be created (`ffd_dll.c`) to run the FFD model. However I couldn't find where the thread is closed. Do you have some insights or can take a look to see whether the code might have some issues?

**Comment by tianwei1989** [comment]
@tsnouidui As far as I know, we did not write any codes to close the thread. When the simulation finishes, the thread is _supposed_ to be closed automatically.

**Comment by jespermattsson** [comment]
Since Dymola runs simulations in a separate process, any threads that are still running when that process ends will be forcibly killed. When simulating an FMU with PyFMI, it will load the DLL in the Python process. In this case the threads will continue to live until they end by themselves or the Python process ends.

The recommended way to handle things like that is an external object. The the Modelica tool is reponsible for ensuring that the constructor is called before the object is used and that the destructor is called when simulation ends.

**Comment by tbeu** [comment]
> As far as I know, we did not write any codes to close the thread.

This really is inappropriate design, that also is appears in Dymola if such models are exported as FMU, as correctly stated by @jespermattsson.

Regarding thread-safety: Please note that variables
- allocated in your co-simulation thread also need to be freed in this thread
- of your co-simulation thread, which are passed (read or write) to/from the Modelica simulation environment need to be thread-synchronized, e.g. by critical sections or mutexes.

**Comment by tianwei1989** [comment]
> As far as I know, we did not write any codes to close the thread.
> This really is inappropriate design, that also is appears in Dymola if such models are exported as FMU, as correctly stated by @jespermattsson.
> 
> Regarding thread-safety: Please note that variables
> 
> allocated in your co-simulation thread also need to be freed in this thread
> of your co-simulation thread, which are passed (read or write) to/from the Modelica simulation environment need to be thread-synchronized, e.g. by critical sections or mutexes.

@tbeu  Thank you for the suggestion. I will have a look at this problem and see if I can do something.

**Comment by mwetter** [comment]
@tianwei1989 : can you please have a look at this issue and submit a pull request that fixes this bug.

**Comment by sewtian** [comment]
> @tianwei1989 : can you please have a look at this issue and submit a pull request that fixes this bug.
Hi Mike, I am looking at the problems. It looks like that I can repeat the problem in JModelica as I simulating the `Buildings.ThermalZones.Detailed.Examples.FFD.ForcedConvection`. I set simulation time as 120 seconds, and the simulation does not stop at 120 seconds. Once I finished the fix, I will send you a pull request.

**Comment by tianwei1989** [comment]
@mwetter I got this error along simulation of `Buildings.ThermalZones.Detailed.Examples.FFD.ForcedConvection` in JModelica in a Windows machine.
![image](https://user-images.githubusercontent.com/5844451/35119368-aad818bc-fc62-11e7-88c2-8d99025f270d.png)

**Comment by mwetter** [comment]
@tianwei1989 Is this something related to your security settings? This message is strange as `cfdReceivedFeedback.c` is not compiled into an `.exe`.

You can try to see if the ground coupled heat exchangers work, which also use C-code from `Buildings/Resources/C-Sources`

**Comment by tianwei1989** [comment]
> @tianwei1989 Is this something related to your security settings? This message is strange as cfdReceivedFeedback.c is not compiled into an .exe.
> 
> You can try to see if the ground coupled heat exchangers work, which also use C-code from Buildings/Resources/C-Sources

I looked into the issue. The reason is possibly because FFD failed to send the feedback command to Modelica at the end of the simulation, as FFD simulations ignores the stop command from the Modelica.

Would you please advise me the exact ground-coupled heat exchanger model to run test on? Thank you.

**Comment by mwetter** [comment]
@tianwei1989 : Try `Buildings.Fluid.HeatExchangers.Ground.Boreholes.BaseClasses.Examples.ExchangeValues` which uses the C function

**Comment by tianwei1989** [comment]
> @tianwei1989 : Try Buildings.Fluid.HeatExchangers.Ground.Boreholes.BaseClasses.Examples.ExchangeValues which uses the C function

@mwetter I did see such warning/error when simulating the aforementioned case. P.S., do you know if there is any graphical interface for JModelica at this moment? I found that OpenModelica is not even good at showing the hierarchy of the `Buildings` library. Would you suggest to graphically check and edit the model in Dymola while running the model in JModelica?

**Comment by tsnouidui** [comment]
I think the` cfdReceivedFeedback.c `error message is generated at compilation not at execution?
Will the model run after you close the window showing that error message?

**Comment by tianwei1989** [comment]
> I think thecfdReceivedFeedback.cerror message is generated at compilation not at execution?
> Will the model run after you close the window showing that error message?

@tsnouidui Yes, it is happened  in the compilation (after I do `compile_fmu()` ). Even I do not close the message, the simulation can run, but not stop at the end of the simulation.
do you know if there is any graphical interface for JModelica at this moment?

**Comment by tsnouidui** [comment]
I don't know of any graphical user interface for JModelica. Best might be to edit the model in Dymola and use JModelica for compilation and simulation.

**Comment by jespermattsson** [comment]
There is no graphical user interface in JModelica.org, only command-line and Python interfaces. It is currently only a compiler and simulation environment, not an editing environment. Sorry.

**Comment by mwetter** [comment]
@tianwei1989 : Regarding "I found that OpenModelica is not even good at showing the hierarchy of the Buildings library.": do you have a reproducable test case that you could document for submission to the OpenModelica development site. My understanding is that OpenModelica's OMEdit does correctly show the hierarchy, but it is in some cases slow, and not all models translate with OpenModelica.

**Comment by sewtian** [comment]
> @tianwei1989 : Regarding "I found that OpenModelica is not even good at showing the hierarchy of the Buildings library.": do you have a reproducable test case that you could document for submission to the OpenModelica development site. My understanding is that OpenModelica's OMEdit does correctly show the hierarchy, but it is in some cases slow, and not all models translate with OpenModelica.

@mwetter it mainly bothers me for its slow speed (for example, I tried to open `Applications.DataCenters.ChillerCooled.Examples.IntegratedPrimaryLoadSideEconomizer`, it is frozen there for more than 20 seconds). I think I might be going to use Dymola to edit the model and JModelica to simulate. Regarding simulating the `Buildings` library in OpenModelica, do you know if the developers have a plan or timeline to make OpenModelica fully support `Modelica.Fluid`?

**Comment by mwetter** [comment]
OpenModelica wants to fully support `Modelica.Fluid` (and `Buildings`) but I don't know about their time line.

**Comment by tianwei1989** [comment]
> @zuowangda JModelica can compile and simulate some of the FFD models
> https://jenkins.jmodelica.org/job/JM-trunk_Buildings-trunk_verify/lastSuccessfulBuild/artifact/diag/index.html. However, some can't be simulated. There seems to be an issue with threads. When I try to simulate an FFD example twice in a row in JModelica, then the first simulation succeeds, whereas the second simulation fails. I looked at bit at the code and saw that a thread will be created (ffd_dll.c) to run the FFD model. However I couldn't find where the thread is closed. Do you have some insights or can take a look to see whether the code might have some issues?

@tsnouidui @zuowangda @sewtian when you say "the first simulation succeeds", do you mean that simulation can run  and STOP at end of the simulation? I added codes to close the thread. However, I found FFD just ignored the "stop" signal send by Modelica. For example, I set simulation time of `Buildings.ThermalZones.Detailed.Examples.FFD.Tutorial.NaturalConvection` case for 60 seconds, however, FFD (does not stop at 60 seconds) ran for 300 seconds, and suddenly collapsed. If this is the case, then I need to understand why FFD ignores stop signal from Modelica. Any idea, or can you please try to reproduce this in Linux?

**Comment by tsnouidui** [comment]
Yes to your first question. What I did was to I write a script to run the same model twice. 
The first run succeeds, the second run fails.

**Comment by tianwei1989** [comment]
> Yes to your first question. What I did was to I write a script to run the same model twice.
> The first run succeeds, the second run fails.

@tsnouidui Below is the script I used to run the case. When the second run fails, where does it fail? Compile? Cannot run? Can run, but won't stop?
`from pymodelica import compile_fmu
from pymodelica import compile_fmu
from pyfmi import load_fmu

model_name='Buildings.ThermalZones.Detailed.Examples.FFD.Tutorial.NaturalConvection'
mo_file = '../modelica-buildings/Buildings'
fmu_model = compile_fmu(model_name,mo_file)

my_fmu = load_fmu(fmu_model)
res = my_fmu.simulate(final_time=60)
`

**Comment by tsnouidui** [comment]
Running the Rooms.FFD.ForcedConvection case twice in a row, I get on the second run

> ERROR in ffd_solver(): Mis-matched synchronization step with t_ffd=40.700000[s], t_cosim=30.000000[s], dt_syn=6.000000[s], dt_ffd=0.100000[s].
> ERROR in para->mytime->t - t_cosim=10.700000
> ERROR in ffd(): FFD solver failed.

**Comment by tianwei1989** [comment]
> Running the Rooms.FFD.ForcedConvection case twice in a row, I get on the second run
> 
> ERROR in ffd_solver(): Mis-matched synchronization step with t_ffd=40.700000[s], t_cosim=30.000000[s], dt_syn=6.000000[s], dt_ffd=0.100000[s].
> ERROR in para->mytime->t - t_cosim=10.700000
> ERROR in ffd(): FFD solver failed.

@tsnouidui In Windows, I am having a different problem than yours. [In my current codes](https://github.com/tianwei1989/modelica-buildings/tree/issue612_ffd_wtian/Buildings), I add the function to close the thread once simulation is done. Would you like to test it in Linux system, while I am working on the issue in Windows? You need to recompile the code.

**Comment by tianwei1989** [comment]
> Since Dymola runs simulations in a separate process, any threads that are still running when that process ends will be forcibly killed. When simulating an FMU with PyFMI, it will load the DLL in the Python process. In this case the threads will continue to live until they end by themselves or the Python process ends.
> 
> The recommended way to handle things like that is an external object. Then the Modelica tool is reponsible for ensuring that the constructor is called before the object is used and that the destructor is called when simulation ends.

@jespermattsson Thank for the information. May I please ask what the relationship between object and constructor/destructor is? Could you please give a more detailed explanation on this? If I add a function (in ffd.dll) to close FFD thread at end of the simulation, what could be the potential risk as compared to the recommended method? Thanks in advance!

**Comment by tsnouidui** [comment]
I did something similar in the past and tried to close the thread as you did with no luck. I am not sure this will solve the issue. I suggest to look at the comment posted by @jespermattsson as well.

**Comment by sewtian** [comment]
> I did something similar in the past and tried to close the thread as you did with no luck. I am not sure this will solve the issue. I suggest to look at the comment posted by @jespermattsson as well.

@tsnouidui Thanks for the feedback, and I have asked @jespermattsson for a more detailed explanation of the suggested method.

**Comment by tianwei1989** [comment]
@tsnouidui A further looking at the compiling information in JModelica ,we found a message interesting:
`The terminal() function-like is not supported, and is currently evaluated to false`.
This might be the reason why FFD cannot stop at end of the simulation, which I found by simulating the case using JModelica in Windows.

**Comment by jespermattsson** [comment]
@tianwei1989 
We have not yet implemented support for the terminal() operator due to lack of use cases. I will bring this use case to Christian's attention.

An external object has associated constructor & destructor functions. The constructor is guaranteed to be called exactly once before the first call to any function taking the external object as an argument (if any such function is called). The destructor is guaranteed to be called exactly once after the last call to any function taking the external object as an argument. This should give a way to start and stop the FFD thread that is well-defined and portable between Modelica tools.

**Comment by sewtian** [comment]
> @tianwei1989
> We have not yet implemented support for the terminal() operator due to lack of use cases. I will bring this use case to Christian's attention.
> 
> An external object has associated constructor & destructor functions. The constructor is guaranteed to be called exactly once before the first call to any function taking the external object as an argument (if any such function is called). The destructor is guaranteed to be called exactly once after the last call to any function taking the external object as an argument. This should give a way to start and stop the FFD thread that is well-defined and portable between Modelica tools.

@jespermattsson Thank you for the answer. My understanding is that we should define the constructor and destructor in a Modelica model. I found some codes like this [online](https://build.openmodelica.org/Documentation/ModelicaReference.Classes.ExternalObject.html),
`class MyTable`
  `extends ExternalObject;`
  `function constructor`
    `input  String  fileName := "";`
    `input  String  tableName := "";`
    `output MyTable table;`
    `external "C" table = initMyTable(fileName, tableName);`
  `end constructor;`

  `function destructor "Release storage of table"`
    `input  MyTable table;`
    `external "C" closeMyTable(table);`
  `end destructor;`
`end MyTable;``

Once I call "MyTable table=MyTable()", it will guaranteed that Table is initialized before making any change to the table, and Table is closed once the simulation is done.

In our case, constructor should be used to created a thread for FFD and destructor should be used to close the thread.

**Comment by jespermattsson** [comment]
@sewtian That is correct, as long as any functions that interact with the table actually takes `table` as an argument. Also note that in the backing C functions, the `table` argument/return value needs to be declared as void pointer. The constructor & destructor will be called once for each instance of MyTable, so in your case you probably want to make sure that there is only a single instance of the external object.

**Comment by sewtian** [comment]
@jespermattsson Thank you. I will start implementing this. Given the fact that JModelica does not support `Terminal()`, my next question is that what can I do to replace `Terminal()` function such that the coupled simulation model can stop at end of the simulation?

**Comment by jespermattsson** [comment]
If you have the coupled simulation as an external object, then the destructor can stop it.

**Comment by sewtian** [comment]
> @sewtian If you have the coupled simulation as an external object, then the destructor can stop it.

@jespermattsson very clear. Thank you.

**Comment by tianwei1989** [comment]
@tsnouidui @jespermattsson implementation is ongoing. The C external function needs to return a float value and a pointer. I am going to put them in a `struct `(defined in a external `.h` file). Just to confirm,  I can declare a variable using this `struct `in Modelica (in a similar way as calling a external c function).

**Comment by jespermattsson** [comment]
> The C external function needs to return a float value and a pointer. I am going to put them in a struct(defined in a external .h file). Just to confirm, I can declare a variable using this structin Modelica (in a similar way as calling a external c function).

@tianwei1989 That depends, what external function are we talking about? If it is the constructor, then no, it has to return only a void pointer. If you need to propagate a float to the Modelica side, then you'll need to add another function that takes the external object as an argument and retrieves the value. (And the float will be passed as a double.)

If it is another function that needs to return two values, then if you declare a Modelica function with multiple outputs and use one of the output variables as an argument to the function in the external clause, then it will be passed as a pointer and the external function is expected to write the result using that pointer.

It's also possible to return a struct. You'll need to declare an equivalent record in Modelica, and the Modelica wrapper function should return an instance of that record. This is however less portable between Modelica tools, as it was previously not very well defined in the Modelica specification. It is also rather limited on what types can be used in the struct (see the Modelica specification). I think that recent JModelica.org versions and recent Dymola versions handle it mostly the same way (except when a record inherits another record), but I'm not sure.

**Comment by sewtian** [comment]
@jespermattsson Yes, I am taking about constructor. To put my question in a more detailed way, here is my code (my intention to implement, and can be very wrong):
`class cfdThread "create constructor and destructor associated with external objects"`
  ` extends ExternalObject;`
  ` // constructor `
  ` function constructor "create ffd.dll or ffd.so"`
  `output Integer retVal`
  ` "Return value of the function (0 indicates CFD successfully started.)";`
  `// FIXME: need to declare a struct FFDThread as void pointer`
  ` output FFDThread FFDThre "the handler of FFD thread";`
  ` // FIXME: need to let cfdStartCosimulation return a construct`
  ` external"C" FFDThre = cfdStartCosimulation();`
  `end constructor;`
`end cfdThread;`

in a `.h` file, I am going to declare the struct `FFDThread` as
`typedef struct {`
`void *PoinFFDThre; `
`double retVal;`
`} FFDThread`

When I call `cfdStartCosimulation`, an external function to create a thread for FFD, I want it to return a double precision `retVal` to show if creating thread is sucessful and `PoinFFDThre` as the pointer to the thread (this will be the input of destrcutor that closes the thread). According to your explain, this is not good.

**Comment by mwetter** [comment]
@sewtian 
Did you look at
`Buildings.Fluid.HeatExchangers.Ground.Boreholes.BaseClasses.Examples.ExchangeValue`?
This function also uses a constructor and a pointer inside a `struct`, similarly to what you need.
Also, if `retVal` is a flag, it should be an integer not a double value. However, you could initialize your pointer to `null` and if creating the thread was successful, then assign the memory to this pointer. This way, you can check the value of the pointer to see whether creating the thread was successful. This is the standard way in C.

**Comment by sewtian** [comment]
> @sewtian
> Did you look at
> Buildings.Fluid.HeatExchangers.Ground.Boreholes.BaseClasses.Examples.ExchangeValue?
> This function also uses a constructor and a pointer inside a struct, similarly to what you need.
> Also, if retVal is a flag, it should be an integer not a double value. However, you could initialize your pointer to null and if creating the thread was successful, then assign the memory to this pointer. This way, you can check the value of the pointer to see whether creating the thread was successful. This is the standard way in C.

@mwetter I looked at the function and it is very similar to the one [online](http://book.xogeny.com/behavior/functions/interpolation/). I will do something like you suggested: just let external C function return the pointer, and check in Modelica if the pointer is `NULL` ; If not, then Modelica knows that creating FFD thread is successful.

**Comment by jespermattsson** [comment]
> [...] and check in Modelica if the pointer is NULL
@sewtian You can't check if the pointer is NULL in Modelica - the language doesn't allow comparisons on pointers. The recommended method for handling failure in external C code is to call the ModelicaError function from the C code. That will have the same effect as a failing assertion in the Modelica code, i.e. in general the simulation will stop with the given error message. If you specifically need a boolean in the model for this, then you'll need to create a C function that takes the pointer as an argument and does the check.

**Comment by mwetter** [comment]
@sewtian Sorry, I meant of course checking in C for `NULL` and if there is an error, call `ModelicaError`

**Comment by sewtian** [comment]
@jespermattsson I created a `CFDThread.mo` file with constructor and destructor that can create and close a thread. Suppose it has a format like this:
`class CFDThread`
  `function constructor`
    `external"C" FFDThre = cfdStartCosimulation(x, y, z...);`
  `end constructor;`
  `function destructor`
    `......`
  `end destructor;`
`end CFDThread;`

Then, I want to instantiate `CFDThread.mo` in `CFDExchange.mo` and use constructor to create thread. Then I would like to do something like this:
`class CFDExchange`
  CFDThread cfdThread = CFDThread(x, y, z...)
  `initial algorithm`
    `x := 1;`
    `y := 1;`
    `z := 1;`
`end CFDExchange;`

Now, my question is will `x`, `y`, and `z` be initialized as `1` before creating the instance `cfdThread (I want my variables to have a specific value  when creating the thread using `constructor`)? Or, in other words, is it possible (legal) to instantiate `CFDThread` after executing the  `initial algorithm` or  `initial equation`?

Thank you in advance for your time.

**Comment by sewtian** [comment]
> @sewtian Sorry, I meant of course checking in C for NULL and if there is an error, call ModelicaError

Thank you Michael. I made some progress and I have a major concern about how to initialize the input before instantiating the Modelica class containing constructor and destrcutor functions. I posted my question and hopefully you and @jespermattsson can give me some clues.

**Comment by jespermattsson** [comment]
@sewtian Since the external object itself is effectively a parameter, all its arguments must be parameters. There is currently a bug that means you do not get an error for assigning parameters from variables, so a constructor call using variables will pass through and just not work properly.

If the parameters used have (fixed = false), then they can be calculated in the initial system as above, as long as the constructor isn't involved in an algebraic loop.

I would advice against using an initial algorithm, as Modelica tools can generally do much more with equations than algorithms. I would only recommend an algorithm if what you need to do is not possible to express in equations.

**Comment by sewtian** [comment]
> @sewtian Since the external object itself is effectively a parameter, all its arguments must be parameters. There is currently a bug that means you do not get an error for assigning parameters from variables, so a constructor call using variables will pass through and just not work properly.
> 
> If the parameters used have (fixed = false), then they can be calculated in the initial system as above, as long as the constructor isn't involved in an algebraic loop.
> 
> I would advice against using an initial algorithm, as Modelica tools can generally do much more with equations than algorithms. I would only recommend an algorithm if what you need to do is not possible to express in equations.

@jespermattsson That answer helps. I believe I know how to work on that part now. Thank you.

**Comment by tianwei1989** [comment]
@mwetter @jespermattsson Currently the implementation is under debugging and testing. Will let you know once they are done.

**Comment by jespermattsson** [comment]
I have just found out that I was wrong about one thing in my last post: Parameters with (fixed = false) can currently _not_ be used as arguments for the constructor. We are currently working on fixing that, though (https://trac.jmodelica.org/ticket/5605).

**Comment by tianwei1989** [comment]
@mwetter @jespermattsson We have been stuck in a problem for a long time. After the model defining constructor and destructor is called in `modelica-buildings/Buildings/ThermalZones/Detailed/BaseClasses/CFDExchange.mo`, the `when sampleTrigger` conditional is never evaluated as true and function `exchange` is not called. By theory, will the calling of a modeling defining constructor and destructor affect the `time` or evaluation of  `sampleTrigger` in a model?

**Comment by tianwei1989** [comment]
Hi @tsnouidui Would you please help me have a look at the reason why in `modelica-buildings/Buildings/ThermalZones/Detailed/BaseClasses/CFDExchange.mo`, the `when sampleTrigger `conditional is never evaluated as true. My branch is [here](https://github.com/tianwei1989/modelica-buildings/tree/issue612_ffd_wtian). You can run on Windows.

**Comment by jespermattsson** [comment]
@tianwei1989 The use of an external object should not affect that. 

Your problems could possibly be caused by https://trac.jmodelica.org/ticket/5234.

I had a brief look at the model, and I have a couple of comments:

When using an external object, unless that object is passed to the external functions using it, then the Modelica tool won't know that the function call depends on the external object. This means that the constructor could be called after the first call to one of the external functions, or the destructor before the last call. If there are no function calls at all that take the external object as parameter, then the tool is not required to call the constuctor & destructor at all. Now with JModelica.org, we do not perform those optimizations, so:
- The external object is always constructed and destructed.
- The destructor is called after the simulation has ended.
- The constructor is called during initialization, of if possible before.
- If the external object is involved in an algebraic loop through its constructor (i.e. a parameter with fixed=false depends on the external object, and the constructor depends on that parameter), then the destructor might be called during initialization so that the object can be re-constructed.
This does mean that if you have a function that depends on the external object having been constructed, and that is called during initialization, then that function must take the external object as an argument, or the calling order might be wrong. With other tools, this might work differently - to be portable, all functions that depend on the external object should take it as a parameter. I think that it should be sufficient to have it as a parameter to the Modelica wrapper, and not pass it along to the C function but that could be open to interpretation.

I can't see why the call to exchange is broken out into its own when-clause and placed in an algorithm. I don't think that would cause any problems, but I'm not sure.

**Comment by xuhan425** [comment]
Dear Dr. Wetter,

We have fixed the ffd issue and sent pull request to issue612_ffd_wtian: #1401 . The changes we made include:

- Using constructor/destructor to send stop command at the end of simulation instead of using terminal() operator. The compile/stop error in JModelica in previous codes was caused by that terminal() operator is currently not supported in JModelica. The revision includes changes in Modelica codes, external "C" and FFD source codes. Now FFD could stop properly in both Dymola and JModelica
- Adding codes in external "C" to avoid exchang() function to be activated more than one times at each data synchronization point, which happens in JModelica
- Adding codes in external "C" to return a "ModelicaError" when multiple FFD instances were created

Just let me know if you have any comments. Thanks for providing so many suggestions to help us on this. @mwetter @tsnouidui @jespermattsson 

Xu @tianwei1989 @zuowangda

**Comment by mwetter** [comment]
@xuhan425: Thanks for your update. I merged it to `issue612_ffd_wtian` and cleaned up the code (in particular checking each `malloc` call for success).

Can you please address the points below after you merged `issue612_ffd_wtian` to your branch.

- [ ] The following line
```
cosim->ffd->msg = (char *) malloc(400*sizeof(char));
```
in cfdStartCosimulation.c seems to assume a maximum string length. Can the FFD code guarantee this length? Better would be to set `msg == NULL` and let the FFD code allocate the string length in order to prevent buffer overflow.

- [ ] In `cfdStartCosimulation.c` there are various calls to `malloc`. These memories need to be freed again when the simulation finishes by calling `free` on all these pointers.

- [ ] Remove all the type conversions of the form
```
... = (REAL) ...
```
As I understand it, REAL is a double, and the type conversion converts a double to a double. Hence I don't see why this is needed.

- [ ] Some code calls `sleep` while other call `Sleep`. If `Sleep` is not needed, then please change all calls to `sleep` (or else use `Sleep` throughout the code).

- [ ] When trying to simulate the examples, I get the error "Component type specifier CFD is a single-instance class. The single-instance must be in the same scope as the class itself."

**Comment by mwetter** [comment]
@xuhan425
- [ ] If `WindowWithoutShade` is changed to have two rooms, then the following error message occurs:
```
The following error was detected at time: 0
ModelicaError: Only one room with FFD can be used, but more than one is used.
The stack of functions is:
Buildings.ThermalZones.Detailed.BaseClasses.CFDThread
Buildings.ThermalZones.Detailed.BaseClasses.CFDThread(3, 0)
Segmentation fault (core dumped)
```
The segmentation fault need to be avoided.

**Comment by xuhan425** [comment]
@mwetter Sure, Dr. Wetter. Will work on this later. Thanks.

**Comment by xuhan425** [comment]
@mwetter 
Dear Dr. Wetter, 

We have fixed the "segmentation fault" and sent you pull request to issue612_ffd_wtian:   #1417. Please have a look and let me know if you have any comments. Thanks.

Xu @tianwei1989  @zuowangda
